### ISSUE TITLE
New field support

### ISSUE THREAD

**Issue by jrock2004** [swastik/ember-inline-edit#2]
Checkbox, select, and select with multi-select

**Comment by swastik** [comment]
Thanks for the issue! :smile: 

If I understand correctly, you mean support for checkbox and select boxes. Is that correct?

If so, any more details on how that would work?

**Comment by jrock2004** [comment]
Sure, for checkboxes

{{ember-inline-edit
    checked=boolean
    field='checkbox'
    onSave=(action "changeValue" "boolean")}}

Just need a flag for if the box should be checked or unchecked. gets its value and send it to the action


For select boxes

{{ember-inline-edit
    multiple=boolean
    options=options
    field='select'
    onSave=(action "changeValue" "selected")}}

For select it can be a little more difficult. Options should be array of options, maybe key value? Also would be nice to support multiple choice. Let me know if this helps

**Comment by swastik** [comment]
Thanks for the details! I'll get around to working on it this week. :smiley:

**Comment by CygnusRoboticus** [comment]
If there hasn't beeen any work done on this yet, I have an idea of how to achieve this: refactor most of the behavior into a `ember-inline-control` mixin, create a separate component `ember-inline-checkbox` and have both that and the existing component extend that. This would also facilitate an `ember-inline-select` component.

The main reason I'd rather have separate components is because right now overriding the template (in order to apply styles/layout) is cumbersome; adding another set of `{{if}}` blocks is going to make that worse (I'd even like to see an `ember-inline-textarea`, actually).

**Comment by swastik** [comment]
That's a good way to go about it. I was wondering whether it would make sense to support custom components (as opposed to just the HTML ones, like text, textarea, select, etc.). A lot of people may be using an ember addon for a select, or maybe even an ember WYSWYG editor, for certain values.

We can allow that with ember's contextual components, like this perhaps:

```
{{#ember-inline-edit isEditing=false as |inline-edit|}}
  {{#inline-edit.editable}}
    Click here to edit
  {{#inline-edit.editable}}

  {{#inline-edit.editor}}
    {{custom-component}}
  {{#inline-edit.editor}}
{{/ember-inline-edit}}
```

This makes is really flexible. It can handle most situations but it also complicates things a bit.

Thoughts?

**Comment by frisco82** [comment]
That would be great, and it will be a much needed component as there are several inline edit or x-editable but there is no way to customize the editable component.
Ember power select does not use contextual components but the architecture design is perfect, maybe you can borrow some ideas.

**Comment by swastik** [comment]
@frisco82 — I'll take a look at power select! 👍

**Comment by swastik** [comment]
Okay, I looked at it! With a select component, the good thing is that it still is very clear with a normal block. Something like:

```
{{#select as |option|}}
   {{option}}
{{else}}
   No options
{{/select}}
```

... reads very well and is extremely clear. For this addon, I got this working:

```
  {{#ember-inline-edit value=value}}
    {{some-custom-component value=value}}
  {{else}}
    <h3>{{value}}</h3>
  {{/ember-inline-edit}}
```

I'm not sure this is very clear though. What do you think?

**Comment by swastik** [comment]
Okay, I gave this a shot. It's on the `block-components` branch. Feel free to take a look.

It'll work something like:

```
  {{#ember-inline-edit as |inline-edit|}}
    {{#inline-edit.editable}}
      <h3>{{value}}</h3>
    {{/inline-edit.editable}}

    {{#inline-edit.editor}}
      {{custom-editor-component value=value}}
    {{/inline-edit.editor}}
  {{/ember-inline-edit}}
```

This is a little verbose but it allows flexibility in both the editor as well as the editable. You can have a select box, for instance, when a user wants to edit, and the value could be wrapped in a heading tag.

We can also make it somewhat less verbose like this:

```
  {{#ember-inline-edit as |inline-edit|}}
    {{#inline-edit.editable value=value}}

    {{#inline-edit.editor}}
      {{custom-editor-component value=value}}
    {{/inline-edit.editor}}
  {{/ember-inline-edit}}
```

Thoughts?

**Comment by CygnusRoboticus** [comment]
That looks like the approach to take, but I think the input/text area logic should be moved to separate components that can provide blocks to ember-inline-edit. You'd have both simpler templates and an example of how to provide blocks.

**Comment by swastik** [comment]
@CygnusRoboticus — Can you give an example? Creating separate components for `text` (and related), `textarea`, and other custom components is fine I think.

**Comment by frisco82** [comment]
May be you can pass a parameter with the component to use for editing as with the value but it could only be used for custom components, It will be very rare that a public component complies with any standard API for value and actions.

That would lead to 3 diferent usages to fit everyone: 

Reduced inline templates
```
  {{#ember-inline-edit as |inline-edit| onSave=(action save)}}
    {{#inline-edit.editable value=value}}
    {{#inline-edit.editor}}
      {{input value=value onBlur=inline-edit.save}}
    {{/inline-edit.editor}}
  {{/ember-inline-edit}}
```
Full inline templates
```
  {{#ember-inline-edit as |inline-edit| onSave=(action save)}}
    {{#inline-edit.editable}}
       <h3 {{action "inline-edit.edit" on="dobleClick"}}>{{value}}<h3>
    {{/inline-edit.editable}}
    {{#inline-edit.editor}}
      {{input value=value onBlur=inline-edit.save}}
    {{/inline-edit.editor}}
  {{/ember-inline-edit}}
```
Custom components
```
  {{#ember-inline-edit as |inline-edit| onSave=(action save)}}
    {{#inline-edit.editable editableComponent='displayValue'}}
    {{#inline-edit.editor editableComponente='editValue'}}
  {{/ember-inline-edit}}
```

**Comment by frisco82** [comment]
In fact maybe the last one could be reduced even more:

`  {{#ember-inline-edit as |inline-edit| onSave=(action save) editableComponent='displayValue' editorComponent='editValue'}}
`

But I am not sure if you cand handle easily optional contextual components

**Comment by swastik** [comment]
@frisco82  — In this case:

```
  {{#ember-inline-edit as |inline-edit| onSave=(action save)}}
    {{#inline-edit.editable editableComponent='displayValue'}}
    {{#inline-edit.editor editableComponente='editValue'}}
  {{/ember-inline-edit}}
```
The custom components will have some data of their own, so you'll end up passing a few more attrs, right? Say you want to pass in a user's email and name, for example. I don't think we can cater to those requirements very well. My thinking is that if there are complex custom components, toggling between them is better as a separate component instead of using an addon like this. It just gives you a lot more control.

For the more common cases like using a custom select box or a custom WYSWYG editor, being able to do that even through an addon like this makes sense.

I like this:

```
  {{#ember-inline-edit as |inline-edit| onSave=(action save)}}
    {{#inline-edit.editable value=value}}
    {{#inline-edit.editor}}
      {{input value=value onBlur=inline-edit.save}}
    {{/inline-edit.editor}}
  {{/ember-inline-edit}}
```

I haven't been able to figure out a way to remove the `editable` component (and just pass in the value) but other than that, I think it's clear enough and can handle the more common cases.

What do you think?

**Comment by frisco82** [comment]
As I tried to say the use case for passing the components directly only makes sense when they are app specific components, but it makes sense if you have a want to reuse the editor or editable markup and not the full component.
For example you can have a editable component that it shared in fields that are checkboxes, textareas and emails.
And the editor maybe it is shared between inline editable and regular fields on different parts of the app.

In those cases you should set an atrribute 'value' on the main ember-inline-edit that is setted to the custom component along with another object with the actions (open, close, save).

**Comment by swastik** [comment]
@frisco82 — That makes sense. 👍 

I'll try to figure out an API for it.

**Comment by CygnusRoboticus** [comment]
@swastik The example I would provide is basically what your last code block is:

  {{#ember-inline-edit as |inline-edit| onSave=(action save)}}
    {{#inline-edit.editable value=value}}
    {{#inline-edit.editor}}
      {{input value=value onBlur=inline-edit.save}}
    {{/inline-edit.editor}}
  {{/ember-inline-edit}}

An `inline-input` and `inline-textarea` component both need templates similar to this, then you can remove the `textFields` and `textAreaFields` logic from `ember-inline-editor`.

It looks like you're already taking a crack at it, but I'm more than willing to contribute if need be.

**Comment by swastik** [comment]
@CygnusRoboticus — Great, I think I get it now! 😄 

I'll work on it on the `block-components` branch this week.

**Comment by swastik** [comment]
Sorry for the (massive) delay!

Largely fixed in #12. 😄
